### ISSUE TITLE
Jenkins kubernetes plugin can't watch pods after 2.2.5 or 2.2.6 upgrade

### ISSUE THREAD

**Issue by ASCIIDuck** [rancher/rancher#21777]
**Issue:**
After upgrading to 2.2.5 or 2.2.6 the Jenkins plugin is no longer able to run builds inside the cluster. It stems from a failure to watch for the newly provisioned pod. The plugin is still able to create and destroy pods but fails on the watch.

Plugin output looks something like this

```
Jul 24, 2019 3:12:24 PM FINE org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud

Building connection to Kubernetes kubernetes URL https://rancher.example.com/k8s/clusters/c-scffx namespace mobilebuilds

Jul 24, 2019 3:12:24 PM FINE org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud

Connected to Kubernetes kubernetes URL https://rancher.example.com/k8s/clusters/c-scffx/

Jul 24, 2019 3:12:24 PM INFO org.csanchez.jenkins.plugins.kubernetes.KubernetesLauncher launch

Created Pod: mobilebuilds/androidbuild-l8zzt

Jul 24, 2019 3:12:24 PM FINE io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager runWatch

Connecting websocket ... io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager@3a01ac53

Jul 24, 2019 3:12:24 PM WARNING io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1 onFailure

Exec Failure: HTTP 403, Status: 403 - null
java.net.ProtocolException: Expected HTTP 101 response but was '403 Forbidden'
	at okhttp3.internal.ws.RealWebSocket.checkResponse(RealWebSocket.java:229)
	at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:196)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:206)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)


Jul 24, 2019 3:12:24 PM FINE io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager scheduleReconnect

Submitting reconnect task to the executor

Jul 24, 2019 3:12:24 PM FINE io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager close

Force closing the watch io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager@3a01ac53

Jul 24, 2019 3:12:24 PM FINE io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$2 execute

Scheduling reconnect task

Jul 24, 2019 3:12:24 PM FINE io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager nextReconnectInterval

Current reconnect backoff is 1000 milliseconds (T0)

Jul 24, 2019 3:12:24 PM WARNING org.csanchez.jenkins.plugins.kubernetes.KubernetesLauncher launch

Error in provisioning; agent=KubernetesSlave name: androidbuild-l8zzt, template=PodTemplate{inheritFrom='', name='AndroidBuild', namespace='mobilebuilds', slaveConnectTimeout=300, idleMinutes=1, label='java8_k', nodeSelector='', nodeUsageMode=EXCLUSIVE, workspaceVolume=EmptyDirWorkspaceVolume [memory=false], containers=[ContainerTemplate{name='jnlp', image='docker.alexandria.example.com/mobilebuild/android-build:1.0.0.8', workingDir='/home/jenkins', command='', args='', ttyEnabled=true, resourceRequestCpu='', resourceRequestMemory='', resourceLimitCpu='', resourceLimitMemory='', livenessProbe=org.csanchez.jenkins.plugins.kubernetes.ContainerLivenessProbe@4d1d8b05}], yamls=[spec:
  containers:
    - name: jnlp
      resources:
         limits:
           memory: "6Gi"
           cpu: "6"]}
io.fabric8.kubernetes.client.KubernetesClientException
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1.onFailure(WatchConnectionManager.java:198)
	at okhttp3.internal.ws.RealWebSocket.failWebSocket(RealWebSocket.java:571)
	at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:198)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:206)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)


Jul 24, 2019 3:12:24 PM FINER org.csanchez.jenkins.plugins.kubernetes.KubernetesLauncher

Removing Jenkins node: androidbuild-l8zzt
```

Entry in the audit log look like this:

```
{
  "auditID": "8bf4d6c2-eeed-40e7-9fef-67a15c4518e5",
  "requestURI": "/k8s/clusters/c-scffx/api/v1/namespaces/mobilebuilds/pods?fieldSelector=metadata.name%3Dandroidbuild-qdb0d&watch=true",
  "user": {},
  "method": "GET",
  "remoteAddr": "10.128.91.212:63935",
  "requestTimestamp": "2019-07-25T17:55:50Z",
  "responseTimestamp": "2019-07-25T17:55:50Z",
  "responseCode": 403,
  "requestHeader": {
    "Accept-Encoding": [
      "gzip"
    ],
    "Connection": [
      "Upgrade"
    ],
    "Origin": [
      "https://rancher.epic.com:-1"
    ],
    "Sec-Websocket-Key": [
      "uhirw9aNFMwpU1rLTUjnAw=="
    ],
    "Sec-Websocket-Version": [
      "13"
    ],
    "Upgrade": [
      "websocket"
    ],
    "User-Agent": [
      "okhttp/3.12.0"
    ]
  },
  "responseHeader": {
    "Content-Type": [
      "application/json"
    ]
  }
}
```

**Steps to reproduce (least amount of steps as possible):**
Upgrade rancher to 2.2.5 or 2.2.6 and use the Jenkin's Kubernetes plugin to run build jobs in the cluster. Happens with the 1.16 and 1.17 versions of the plugin. The 403s occur even if the user being used by Jenkins is a rancher-wide admin.

**Comment by ASCIIDuck** [comment]
#21740 may be related

**Comment by patrick0057** [comment]
I have tested this in a lab and also had this issue on 2.2.5 and 2.2.6.  Problem does not exist on 2.2.4.

**Comment by alena1108** [comment]
@patrick0057 have you tested it with the same version of k8s in 2.2.5 and 2.2.4?

**Comment by patrick0057** [comment]
To clarify, I setup a Rancher v2.2.6 with Kubernetes 1.13.5 and then launched Jenkins from helm repository.  I setup a project to run a bash script that ended with a 403 forbidden.

**Comment by cjellick** [comment]
@kjcjohnson has hit the nail on the head with the issue and PR they opened here https://github.com/fabric8io/kubernetes-client/issues/1667 and here https://github.com/fabric8io/kubernetes-client/pull/1669

We added Origin checking in v2.2.5 to prevent Cross-Site Websocket Hijacking (see https://www.christian-schneider.net/CrossSiteWebSocketHijacking.html for an explanation). The client is sending a malformed Origin header and is thus triggering our logic that rejects the requests as it appears to be coming from a different origin.

There might be a workaround until the fabric8io client is fixed where the `serverUrl` in the `jenkins-jenkins-kubernetes-config` configMap is set **with** a port so that the client doesnt set it to -1 (ie like `https://my.rancher:443/k8s/local`), but I haven't yet found how that configMap is getting generated or if simply modifying it after the fact would alleviate the issue.

If anyone is interested in trying, please take a look at that configMap (in the same namespace as the jenkins app) and add port 443 to the serverUrl field.

**Comment by cjellick** [comment]
Per @vincent99:
We could technically apply the check only if another browser-y header is present (like `User-Agent` contains `mozilla`), since you can't change that in a browser websocket call and if you're anywhere else you'll be able to set arbitrary headers to arbitrary values (like `Origin`)


This might be the fix we go with so that we don’t have to wait for the client fix to get merged and trickle into a release or jenkins

**Comment by bmdepesa** [comment]
Reproduced in rancher/rancher:v2.2.6

* Create cluster
* Enable Helm catalog
* Create jenkins-custom secret w/ access token
* Make sure serverUrl is edited in YAML file contents
* Launch Jenkins app in cluster
  * Edit yaml and paste contents from file
* Create ingress to Jenkins workload
* Retrieve Jenkins admin password from secret
* Access Jenkins
* _In my case I had to remove the memory limits for the Kubernetes pod template in Jenkins system configuration_*
* Create job w/ sh build step & execute
* Error shown in Jenkins workload logs


```
# *********** Notes *************
# This file relies on the following variables being passed as Kubernetes secrets
# Create a secret called "jenkins-custom"
#   Add the following field to it
#   * JENKINS_BUILDER_SECRET:  This is a kubectl user key or Ranhcer API key that has the ability to manage the jenkins namespace.   We use the API key of a user in Rancher, format should look like the following, where kubeconfig is the name of our custom user and u-5nrwf is the ID of the user, this is displayed in Rancher when you create an API key.
#     kubeconfig-u-3rcbs:<long-secret-text-here>
persistence:
   enabled: false
master:
  componentName: "jenkins-master"
  image: "jenkins/jenkins"
  tag: "lts-alpine"
  imagePullPolicy: "IfNotPresent"
  servicePort: 8080
  serviceType: ClusterIP
  containerEnv:
    - name: CASC_JENKINS_CONFIG
      value: "/var/jenkins_home/casc_configs/..data/jenkins.yaml"
    - name: SECRETS
      value: "/var/jenkins_secrets"
  cpu: "1000m"
  memory: "4096Mi"
  secretsFilesSecret: "jenkins-custom"
  JCasC:
    enabled: true
    configScripts:
      credentials-config: |
        credentials:
          system:
            domainCredentials:
            - credentials:
              - string:
                  description: "Rancher Jenkins Builder"
                  id: "jenkins-builder"
                  scope: GLOBAL
                  secret: ${JENKINS_BUILDER_SECRET}
      kubernetes: |
        jenkins:
          clouds:
            - kubernetes:
                name: "test-cluster"
                # Rancher latest cluster
                serverUrl: "<URL>"
                skipTlsVerify: true
                namespace: "jenkins"
                credentialsId: "jenkins-builder"
                jenkinsUrl: "http://jenkins:8080"
                jenkinsTunnel: "jenkins-agent:50000"
                connectTimeout: 0
                readTimeout: 0
                containerCapStr: "100"
                maxRequestsPerHostStr: "64"
                retentionTimeout: 5
                templates:
                  - name: "jenkins-builder"
                    namespace: "jenkins"
                    label: "linux"
                    nodeUsageMode: EXCLUSIVE
                    idleMinutes: 1
                    idleMinutesStr: "1"
                    activeDeadlineSeconds: 120
                    activeDeadlineSecondsStr: "120"
                    slaveConnectTimeout: 1000
                    slaveConnectTimeoutStr: "1000"
                    containers:
                      - name: "jnlp"
                        command: ''
                        args: ''
                        image: "jenkins/jnlp-slave:alpine"
                        alwaysPullImage: false
                        workingDir: "/home/jenkins"
                        ttyEnabled: true
                        resourceRequestCpu: "200m"
                        resourceLimitCpu: "500m"
                        resourceRequestMemory: "256Mi"
                        resourceLimitMemory: "512Mi"
                    volumes:
                      - emptyDirVolume:
                          memory: false
                          mountPath: "/tmp"
  sidecars:
    configAutoReload:
      enabled: true
      resources:
        limits:
          cpu: 100m
          memory: 100Mi
        requests:
          cpu: 50m
          memory: 50Mi
      sshTcpPort: 1044
      label: jenkins_config
      folder: /var/jenkins_home/casc_configs
  installPlugins:
    - configuration-as-code:latest
    - configuration-as-code-support:latest
    - ansicolor:latest
    - workflow-multibranch:latest
    - kubernetes:latest
    - kubernetes-credentials:latest
    - workflow-aggregator:latest
    - workflow-job:latest
    - credentials-binding:latest
    - blueocean:latest  
agent:
  enabled: false
```

**Comment by bmdepesa** [comment]
rancher/rancher:master-head bc3700f8

* Reproduced in v2.2.6
* Deploy rancher/rancher:master-head
* Create cluster
* Enable Helm catalog
* Create jenkins-custom secret w/ access token
* Make sure serverUrl is edited in YAML file contents
* Launch Jenkins app in cluster
  * Edit yaml and paste contents from file
* Create ingress to Jenkins workload
* Retrieve Jenkins admin password from secret
* Access Jenkins
* Create job w/ sh build step
* Job builds without error

**Comment by jonesbusy** [comment]
Hi!

Same issue after upgrade from rancher 2.2.7 (K8S 1.15.5) to 2.2.8 (K8S 1.15.6).

The Jenkins Kubernetes Plugin stop working.

We had to rollback and cannot update K8S and have latest security fixes

Is it an issue on Rancher side ?

```
WARNING: Exec Failure: HTTP 403, Status: 403 -
java.net.ProtocolException: Expected HTTP 101 response but was '403 Forbidden'
at okhttp3.internal.ws.RealWebSocket.checkResponse(RealWebSocket.java:229)
at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:196)
at okhttp3.RealCall$AsyncCall.execute(RealCall.java:206)
at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
```

Regards,

**Comment by cprivite** [comment]
The workaround mentioned by @cjellick (from the @kjcjohnson issue #21740 ) is still working for us on 2.2.8.

**Comment by jonesbusy** [comment]
Ok thanks, I miss it. It seems that the workaround work also for us (adding 443 port on the URL).

**Comment by jonesbusy** [comment]
Fixed on 1.18.2 of the Jenkins plugin : https://github.com/jenkinsci/kubernetes-plugin/releases

Workaround to add the port is not necessary anymore

**Comment by cjellick** [comment]
@rmweir can you get this updated in rancher? I _think_ you just need to update here: https://github.com/rancher/types/blob/master/apis/management.cattle.io/v3/tools_system_images.go#L22
cc @loganhz @gitlawr

**Comment by cjellick** [comment]
I have a release for it: https://github.com/rancher/pipeline-jenkins-server/releases/tag/v0.1.4

**Comment by bmdepesa** [comment]
rancher/rancher:master-head 57508ebba

* With the updated k8s plugin, I am not able to reproduce this issue. 
* Create cluster
* Enable pipeline
* Pipeline executes successfully

additionally, installed Jenkins as an app was able to create executors and complete successfully as well.

**Comment by erenatas** [comment]
> Fixed on 1.18.2 of the Jenkins K8S plugin : https://github.com/jenkinsci/kubernetes-plugin/releases
> 
> Workaround to add the port is not necessary anymore

Thank you so much!
